Source/DafnyCore/AST/AstVisitor.cs

```diff
@@ -0,0 +1,333 @@
+        VisitStatement(method.Body, context);
+      }
+    }
+
+    private void VisitDefaultParameterValues(List<Formal> formals, VisitorContext context) {
+      formals
+        .Where(formal => formal.DefaultValue != null)
+        .Iter(formal => VisitExpression(formal.DefaultValue, context));
+    }
+
+    private void VisitAttributedExpression(AttributedExpression attributedExpression, VisitorContext context) {
+      VisitAttributes(attributedExpression, context.EnclosingModule);
+      VisitExpression(attributedExpression.E, context);
+    }
+
+    private void VisitAttributes(IAttributeBearingDeclaration parent, ModuleDefinition enclosingModule) {
+      foreach (var attribute in parent.Attributes.AsEnumerable()) {
+        foreach (var arg in attribute.Args ?? Enumerable.Empty<Expression>()) {
+          VisitExpression(arg, GetContext(new NoContext(enclosingModule), false));
+        }
+      }
+    }
+
+    protected virtual void VisitUserProvidedType(Type type, VisitorContext context) {
+    }
+
+    protected virtual void VisitExpression(Expression expr, VisitorContext context) {
+      if (VisitOneExpression(expr, context)) {
+        // Visit user-provided types
+        if (expr is SeqConstructionExpr seqConstructionExpr) {
+          var userProvidedElementType = seqConstructionExpr.ExplicitElementType;
+          if (userProvidedElementType != null) {
+            VisitUserProvidedType(userProvidedElementType, context);
+          }
+
+        } else if (expr is TypeUnaryExpr typeUnaryExpr) {
+          VisitUserProvidedType(typeUnaryExpr.ToType, context);
+
+        } else if (expr is LetExpr letExpr) {
+          foreach (var lhs in letExpr.LHSs) {
+            foreach (var v in lhs.Vars) {
+              VisitUserProvidedType(v.SyntacticType, context);
+            }
+          }
+
+        } else if (expr is QuantifierExpr quantifierExpr) {
+          foreach (BoundVar v in quantifierExpr.BoundVars) {
+            VisitUserProvidedType(v.Type, context);
+          }
+
+        } else if (expr is SetComprehension setComprehension) {
+          foreach (BoundVar v in setComprehension.BoundVars) {
+            VisitUserProvidedType(v.Type, context);
+          }
+
+        } else if (expr is MapComprehension mapComprehension) {
+          foreach (BoundVar v in mapComprehension.BoundVars) {
+            VisitUserProvidedType(v.Type, context);
+          }
+
+        } else if (expr is LambdaExpr lambdaExpr) {
+          foreach (BoundVar v in lambdaExpr.BoundVars) {
+            VisitUserProvidedType(v.Type, context);
+          }
+
+        } else if (expr is MatchExpr matchExpr) {
+          foreach (MatchCaseExpr mc in matchExpr.Cases) {
+            foreach (BoundVar v in mc.Arguments) {
+              VisitUserProvidedType(v.Type, context);
+            }
+          }
+        }
+
+        // Visit subexpressions
+        expr.SubExpressions.Iter(ee => VisitExpression(ee, context));
+
+        // Visit substatements
```

Comment:
Can you please visit substatements before subexpressions? Since all our substatements are before expressions such as lemma calls and assertions, I would prefer the visit to be in chronological order.